### ISSUE TITLE
Mithril 1.0.1 does not write "value" attributes for input elements to the DOM

### ISSUE THREAD

**Issue by kyleolivo** [lhorie/mithril.js#1740]
The value attribute is not being set in the DOM when drawing an input element:

```javascript
m("input", {type: "text", value: "foo"}) // outputs <input type="text"/> only
```

See here for an example: https://jsbin.com/cokezaxoxi/edit?html,js,output

**Comment by tivac** [comment]
Chrome Stable
![image](https://cloud.githubusercontent.com/assets/49545/24310067/b6cceff8-108b-11e7-825b-afa21ed7a3ad.png)

Firefox Dev Edition
![image](https://cloud.githubusercontent.com/assets/49545/24310089/c336b6ac-108b-11e7-84b6-71a31e350486.png)

Edge
![image](https://cloud.githubusercontent.com/assets/49545/24310099/cd034498-108b-11e7-8c35-9ab1040c0cef.png)

IE 11
![image](https://cloud.githubusercontent.com/assets/49545/24310111/d77481d0-108b-11e7-8244-58e525d4e2b1.png)

It seems like there's something else going on here, I can't repro anywhere. Can you give some more details on the browser you're seeing this in?

**Comment by pygy** [comment]
It works in Safari 10.0.3 as well...

`value` is set as a property (conceptually `vnode.dom.value = '...'`, not `vnode.dom.setAttribute('value', '...')`), it isn't reflected as an attribute in the DOM inspector. Nevertheless, it works fine.

**Comment by marques-work** [comment]
I was able to reproduce it in Chrome 56.0.2924.87 (Mac OS X). It does display in the UI, but as @pygy demonstrated, it isn't written as a DOM attribute. Setting the value property does work, but it's unexpected behavior to not output the attribute to the DOM, being that the signature (http://mithril.js.org/hyperscript.html) is:

```javascript
vnode = m(selector, attributes, children) // attributes is described as "HTML attributes"
```

**Comment by barneycarroll** [comment]
Mithril's `attributes` does not map directly to DOM attributes. `value` is a (like `checked`, `scrollTop`, and others) is a DOM *property* and as such is set directly as an object property of the element rather than as an attribute. When writing plain HTML there is no distinction — you must provide initial properties in the same way you provide attribute. But because Mithril operates in a live DOM it must make the distinction. 

To put it another way, setting `value` as a DOM attribute would not have an effect on the dynamic behaviour of the `value` property.

There are some special exceptions: custom elements (native web components) require all input to be set as DOM attributes in order to trigger their internal lifecycles. But in this case, it isn't necessary.

**Comment by marques-work** [comment]
@barneycarroll another exception is <input type="reset" />; Without an initial DOM attribute value, this control will clear the fields instead of setting them to the initial value in the DOM. There are other workarounds of course, but that's another valid case.

**Comment by pygy** [comment]
@marques-work The engine already takes many corner cases into account, see https://github.com/lhorie/mithril.js/blob/ad144072b622f063dcd1c7757978302b4c7ca492/render/render.js#L468-L498 for the current code

**Comment by marques-work** [comment]
@pygy thanks for the context. I think @kyleolivo and I are fine with closing the issue as the behavior was an explicit implementation decision. It just caught us off guard that value was treated differently from the other common HTML attributes on input elements.

**Comment by pygy** [comment]
@marques-work You're welcome :-)
